Sync transactions with exchanger maksym khomishchak

Max-Homishchak opened the pull request.

mkhomishchak committed:
Added deposit withdrawal transaction sync and caching

added new API for sync with app to date transactions data
added transactional removal of transactions cache after balance removal
update naming
repackaging

mkhomishchak committed:
Add remote type for balance operations service

add remote type for balance operation service
fix sync issue with saving the same balance once again with the same balance id

mkhomishchak committed:
Merge branch 'dev' into sync-transactions-with-exchanger_Maksym-Khomishchak

maxmsi97 commented:
LGTM!

kbabikAmelco reviewed (approved):
LGTM!

DendeberiaOleksandr commented on src/main/java/com/khomishchak/cryptoportfolio/controllers/ExchangerController.java:
```diff
@@ -46,22 +48,25 @@ public List<Balance> getAccountBalances(@RequestAttribute long userId) {
-        return exchangerService.deleteExchangerForUser(userId, ExchangerCode.valueOf(exchangerCode));
+    @DeleteMapping("/{balanceId}/balance")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void deleteAccountBalance(@PathVariable long balanceId) {
```
Why do you have account's balance related logic inside exchanger controller. From perspective of reading this controller should provide API to manage exchanger resource. Please explain how does an exchanger as an entity handle account's balance and what is a relation between them.

mkhomishchak replied:
this controller is responsible for everything that is related to the 3rd party exchangers, and this balance is a part of exact exchanger(each exchanger integration have its own balance), and so balance is isolated exchanger resource

DendeberiaOleksandr replied:
Path variable means a path to current resource. In your case it's `Exchanger` resource so path variable should be some identifier of exchanger, e.g id, unique name etc. If exchanger has list of balances then it should looks like:
```suggestion
    @DeleteMapping("/balances/{balanceId}")
```

mkhomishchak replied:
for now it is not required change. For now one exchanger for one user can have only one balance

DendeberiaOleksandr replied:
@mkhomishchak If I got it there is no `Exchanger` entity itself. The representation if it is a `Balance` entity and each account may have 0 or more balances. If it's why isn't better to have a user's API what handles users' resources. Then it will be more intuitive `/users/{userId}/balances/{balanceId} `

DendeberiaOleksandr commented on src/main/java/com/khomishchak/cryptoportfolio/controllers/ExchangerController.java:
```diff
@@ -46,22 +48,25 @@ public List<Balance> getAccountBalances(@RequestAttribute long userId) {
-    public List<DepositWithdrawalTransaction> getAccountWithdrawalDepositWalletHistory(@RequestAttribute long userId,
-            @PathVariable String exchangerCode) {
-        return exchangerService.getWithdrawalDepositWalletHistory(userId, ExchangerCode.valueOf(exchangerCode));
+    @GetMapping("/deposit-withdrawal-history")
```
Same as above. From reader perspective and constructing an API this endpoint returns deposit withdrawal history of all exchangers for some user id. Is that right?

mkhomishchak replied:
yes, that's right, because this history is also related to each of the exchangers integrations. there no sense to create a new controller with just 2-3 methods related to deposit withdrawal history

mkhomishchak committed:
Merge branch 'dev' into sync-transactions-with-exchanger_Maksym-Khomishchak

mkhomishchak committed:
add error handling, delete dw transactions after delete of balance

throw BalanceNotFound when we do not have integration for userId and Code, worth to add validator before integration layer
delete dw transactions not only in the cache, but also in the db when balance was deleted

DendeberiaOleksandr commented on src/main/java/com/khomishchak/ws/controllers/ExchangerController.java:
```diff
@@ -44,22 +45,25 @@ public List<Balance> getAccountBalances(@RequestAttribute long userId) {
-    @DeleteMapping("/{exchangerCode}/balance")
-    public DeleteExchangerResp deleteAccountBalance(@RequestAttribute long userId, @PathVariable String exchangerCode) {
-        return exchangerService.deleteExchangerForUser(userId, ExchangerCode.valueOf(exchangerCode));
+    @DeleteMapping("/{balanceId}/balance")
```
Incorrect path name.
```suggestion
    @DeleteMapping("/balances/{balanceId}")
```

DendeberiaOleksandr commented on src/main/java/com/khomishchak/ws/controllers/ExchangerController.java:
```diff
@@ -44,22 +45,25 @@ public List<Balance> getAccountBalances(@RequestAttribute long userId) {
     }
 
-    // TODO: create sync API for all exchanger related data
+    @PostMapping("/synchronize/deposit-withdrawal-history")
```
```suggestion
    @PostMapping("/deposit-withdrawal-history/synchronize")
```

Max-Homishchak committed:
updated API
